internal/kafka/internal/services/data_plane_kafka.go

```diff
@@ -52,67 +64,126 @@ func NewDataPlaneKafkaService(kafkaSrv KafkaService, clusterSrv ClusterService,
-			e = d.persistKafkaRoutes(kafka, ks, cluster)
-		case statusError:
-			// when getStatus returns statusError we know that the ready
-			// condition will be there so there's no need to check for it
-			readyCondition, _ := ks.GetReadyCondition()
-			e = d.setKafkaClusterFailed(kafka, readyCondition.Message)
-		case statusDeleted:
-			e = d.setKafkaClusterDeleting(kafka)
-		case statusRejected:
-			e = d.reassignKafkaCluster(kafka)
-		case statusRejectedClusterFull:
-			e = d.unassignKafkaFromDataplaneCluster(kafka)
-		case statusUnknown:
-			log.Infof("kafka cluster %s status is unknown", ks.KafkaClusterId)
-		default:
-			log.V(5).Infof("kafka cluster %s is still installing", ks.KafkaClusterId)
-		}
-		if e != nil {
-			log.Error(errors.Wrapf(e, "Error updating kafka %s status", ks.KafkaClusterId))
+
+	prewarmingStatusInfo := map[string]map[string]int{}
+	supportedInstanceTypes := cluster.GetSupportedInstanceTypes()
+
+	// prefill prewarming status info with empty count
+	for _, supportedInstanceType := range supportedInstanceTypes {
+		// sets all the statuses to 0 to allow for the count to drop
+		prewarmingStatusInfo[supportedInstanceType] = map[string]int{
+			statusReady.String():      0,
+			statusError.String():      0,
+			statusUnknown.String():    0,
+			statusInstalling.String(): 0,
 		}
+	}
 
-		e = d.setKafkaRequestVersionFields(kafka, ks)
-		if e != nil {
-			log.Error(errors.Wrapf(e, "Error updating kafka '%s' version fields", ks.KafkaClusterId))
+	for _, ks := range status {
+		managedKafkaDeploymentType := getManagedKafkaDeploymentType(ks)
+		switch managedKafkaDeploymentType {
+		case realDeploymentType:
+			d.processRealKafkaDeployment(ks, cluster, log)
+		case reservedDeploymentType:
+			d.processReservedKafkaDeployment(ks, prewarmingStatusInfo, log, clusterID)
 		}
 	}
+
+	// updates the cluster prewarming status count
+	cluster.SetPrewarmingStatusInfo(prewarmingStatusInfo)
+	err = d.clusterService.Update(*cluster)
+	if err != nil {
+		log.Error(err)
+	}
+
 	return nil
 }
 
+// processRealKafkaDeployment process reserved kafka instances and updates their status in cluster
+func (d *dataPlaneKafkaService) processReservedKafkaDeployment(ks *dbapi.DataPlaneKafkaStatus, prewarmingStatusInfo map[string]map[string]int, log logger.UHCLogger, clusterID string) {
+	parsedID := strings.Split(ks.KafkaClusterId, "-")
+	if len(parsedID) < 3 {
+		log.Error(fmt.Errorf("The reserved %q ID does not follow the format 'reserved-kafka-<instance-type>-<number>'", ks.KafkaClusterId))
+		return
+	}
+
+	instanceType := parsedID[2]
+	_, ok := prewarmingStatusInfo[instanceType]
+	if !ok {
+		log.Error(fmt.Errorf("The reserved %q ID is not supported in the cluster with cluster_id %q", ks.KafkaClusterId, clusterID))
+		return
+	}
+
+	s := getStatus(ks)
+
+	if s != statusReady && s != statusInstalling && s != statusError {
+		// consider the rest of the statuses as 'unknown' in the prewarming context as they shouldn't be sent by the Operator
```

Comment:
Hmm actually, it is just the these two statuses that we are ignoring.
It does not hurt to add and initialize them to `0`. If the Fleetshard starts sending them, we'll have the count anyway. Otherwise, they'll be set to `0` always something that won't hurt the metric in any way.